src/kokkos/plugin-PixelTriplets/kokkos/CAHitNtupletGeneratorKernels.cc

```diff
@@ -34,116 +34,166 @@ namespace KOKKOS_NAMESPACE {
     // applying conbinatoric cleaning such as fishbone at this stage is too expensive
     //
 
-#ifdef TODO
-    auto nthTot = 64;
-    auto stride = 4;
-    auto blockSize = nthTot / stride;
-    auto numberOfBlocks = (3 * m_params.maxNumberOfDoublets_ / 4 + blockSize - 1) / blockSize;
-    auto rescale = numberOfBlocks / 65536;
-    blockSize *= (rescale + 1);
-    numberOfBlocks = (3 * m_params.maxNumberOfDoublets_ / 4 + blockSize - 1) / blockSize;
-    assert(numberOfBlocks < 65536);
-    assert(blockSize > 0 && 0 == blockSize % 16);
-    dim3 blks(1, numberOfBlocks, 1);
-    dim3 thrs(stride, blockSize, 1);
-
-    kernel_connect<<<blks, thrs, 0, cudaStream>>>(
-        device_hitTuple_apc_,
-        device_hitToTuple_apc_,  // needed only to be reset, ready for next kernel
-        hh.view(),
-        device_theCells_.get(),
-        device_nCells_,
-        device_theCellNeighbors_,
-        device_isOuterHitOfCell_.get(),
-        m_params.hardCurvCut_,
-        m_params.ptmin_,
-        m_params.CAThetaCutBarrel_,
-        m_params.CAThetaCutForward_,
-        m_params.dcaCutInnerTriplet_,
-        m_params.dcaCutOuterTriplet_);
-    cudaCheck(cudaGetLastError());
+    int nthTot = 64;
+    int stride = 4;
+    int teamSize = nthTot / stride;
+    int leagueSize = (3 * m_params.maxNumberOfDoublets_ / 4 + teamSize - 1) / teamSize;
+    int rescale = leagueSize / 65536;
+    teamSize *= (rescale + 1);
+    leagueSize = (3 * m_params.maxNumberOfDoublets_ / 4 + teamSize - 1) / teamSize;
+    assert(leagueSize < 65536);
+    assert(teamSize > 0 && 0 == teamSize % 16);
+    teamSize *= stride;
+
+    // Current multi-team algorithms are usually based on CUDA implementations which
+    // may not be flexible enough to become a generic parallel solution for all backends.
+    // Thus team policy should be manually handled for each specific backend.
+#ifdef KOKKOS_BACKEND_CUDA
+    Kokkos::TeamPolicy<KokkosExecSpace> policy{execSpace, leagueSize, teamSize};
+#else  // serial
+    Kokkos::TeamPolicy<KokkosExecSpace> policy{execSpace, 1, Kokkos::AUTO()};
+#endif
+    const auto *hhp = hh.view();
+    Kokkos::parallel_for(
+        "kernel_connect", policy, KOKKOS_LAMBDA(const Kokkos::TeamPolicy<KokkosExecSpace>::member_type &teamMember) {
+          kernel_connect(device_hitTuple_apc_,
+                         device_hitToTuple_apc_,  // needed only to be reset, ready for next kernel
```

Comment:
I went ahead and started testing running with this PR. One thing I noticed is that passing member `Kokkos::View`s leads to `an illegal memory access was encountered` error on CUDA, because `this` pointer is in the host memory. So all member `Kokkos::View`s need to be copied to local `Kokkos::View`s before the `parallel_for`. 

I see we actually "knew" this already :)
https://github.com/cms-patatrack/pixeltrack-standalone/blob/1593be6653244785747eef0184f5807c3a46b6a2/src/kokkos/plugin-SiPixelClusterizer/kokkos/SiPixelRawToClusterGPUKernel.cc#L594-L601

Could you change that for all kernel launches in this PR? I can take care of the rest.

Reply:
Done.